### PR TITLE
Intersphinx object inventories don't load properly into dotnet domain

### PR DESCRIPTION
**Pull request by agjohnson**
Intersphinx mappings load inventories into the domain incorrectly. Test this more.

### CODE REVIEW

File: sphinxcontrib/dotnetdomain.py [hunk 1]
```diff
@@ -571,6 +573,23 @@ class DotNetDomain(Domain):
...
+        for name, obj in iteritems(self.object_types):
+            for roles in obj.roles:
+                if isinstance(roles, string_types):
+                    roles = [roles]
+                try:
+                    for rolename in roles:
+                        self._role2type.setdefault(rolename, []).append(name)
+                except TypeError:
+                    pass
+            self._type2role[name] = obj.roles[0] if obj.roles else ''
```
Comment (ericholscher) [inline sphinxcontrib/dotnetdomain.py]: Will this not be an issue when obj.roles is a `string_type`, which the check above is looking for?

Comment (agjohnson) [inline sphinxcontrib/dotnetdomain.py]: Missed this, yeah it would

**Comment by ericholscher** [comment]
LGTM, though I'm not sure about it solving the actual problem. I'm guessing testing this is hard? We could perhaps generate the intersphinx inventory and then process it to test it round trips properly?

**Comment by agjohnson** [comment]
Yeah, was not sure how to best write a test for this. Perhaps mocking the intersphinx load function and testing a link is resolved? Seems doable

**Comment by ericholscher** [comment]
You could probably run the actual code and validate it -- https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/intersphinx.py#L79

**Comment by agjohnson** [comment]
Yeah, though I'd probably just skip that step to avoid mucking around with their compressed inventory file format, it's a pain. Mocked an actual object would be much easier

**Comment by agjohnson** [comment]
Tests attached. I wrote a mocked XML builder to provide XML string output. The text output doesn't actually output a linked syntax, only the literal that's quoted. The XML outputs a full link that we can inspect actually resolved.

**Comment by ericholscher** [comment]
:+1: Looks good

**Pull request merged**